src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala

```diff
@@ -0,0 +1,47 @@
+package org.broadinstitute.dsde.workbench.sam
+package backgroundProcess
+
+import cats.effect.{ContextShift, IO}
+import fs2.Stream
+import io.chrisdavenport.log4cats.StructuredLogger
+import io.circe.Decoder
+import org.broadinstitute.dsde.workbench.google2.{Event, GoogleSubscriber}
+import org.broadinstitute.dsde.workbench.model.WorkbenchUserId
+import org.broadinstitute.dsde.workbench.sam.service.UserService
+import org.broadinstitute.dsde.workbench.sam.util.SamRequestContext
+
+// Each instance of Sam will run a subscriber, since they all subscribe to the same topic via same subscription,
+// messages will be distributed among subscribers instead of distributed to all of them, which is what we want.
+class CryptominingSubscriber(subscriber: GoogleSubscriber[IO, CryptominingUserMessage],
+                             userService: UserService
+                            )(implicit logger: StructuredLogger[IO], cs: ContextShift[IO]) {
+  val process: Stream[IO, Unit] = subscriber.messages
+    .parEvalMapUnordered(2)(messageHandler) // concurrency is low since we don't need high performance in disabling cryptomining users.
+    .handleErrorWith(error => Stream.eval(logger.error(error)("Failed to initialize message processor")))
+
+  private def messageHandler(event: Event[CryptominingUserMessage]): IO[Unit] = {
+    for {
+      res <- messageResponder(event.msg).attempt
+      _ <- res match {
+        case Left(e)  => logger.error(e)("Fail to process pubsub message")
+        case Right(_) => IO.unit
+      }
+      // acking the message whether or not we successfully processed the message since this is not essential part of
+      // terra functionality. That is, if we didn't disable a bad user, it's no big deal.
+      // To conditionally acking message, we'll need to config a dead letter topic where messages reach max retry can go to,
+      // which we're not doing for this use case.
+    _ <- IO{event.consumer.ack()}
+    } yield ()
+  }
+
+  private def messageResponder(
+                                         message: CryptominingUserMessage
+                                       ): IO[Unit] = IO.fromFuture(IO(userService.disableUser(message.userSubjectId, SamRequestContext(None)))).void
+}
```

Comment:
what is the test in your mind trying to validate? if it's just disabling user, I think it should be covered in userService's unit test

Reply:
that makes sense. Would it be useful to test this subscriber? (I don't know how we would test it either, but just wanted to check and ask to understand why we would or wouldn't want to test this)

Reply:
anything specific you want to test about the subscriber?

Subscriber has unit test in google2 to make sure it can receive message properly.
The only thing worth testing here is the actual handler, but the handler here is quite simple cuz it's just calling userSerivce.disableUser, which I assume is covered by userSerivce's unit test